libzmix/src/verifiable_encryption/cs_verifiable_encryption.rs

```diff
@@ -0,0 +1,849 @@
+// Copyright contributors to Hyperledger Ursa.
+// SPDX-License-Identifier: Apache-2.0
+
+/// Camenisch-Shoup verifiable encryption.
+/// Based on the paper Practical Verifiable Encryption and Decryption of Discrete Logarithms. https://www.shoup.net/papers/verenc.pdf
+/// Need to be used with Anonymous credentials as described in Specification of the Identity
+/// Mixer Cryptographic Library, https://domino.research.ibm.com/library/cyberdig.nsf/papers/EEB54FF3B91C1D648525759B004FBBB1/$File/rz3730_revised.pdf
+/// sections 5.3, 6.2.10 and 6.2.19
+use super::bn::{BigNumber, BigNumberContext, BIGNUMBER_1, BIGNUMBER_2};
+use super::errors::prelude::*;
+
+use super::cl::constants::*;
+use super::cl::hash::get_hash_as_int;
+use super::cl::helpers::*;
+
+#[derive(Serialize, Deserialize)]
+pub struct PaillierGroup {
+    pub g: BigNumber,
+    pub h: BigNumber,
+    pub n_by_4: BigNumber, // n/4, precomputation
+    pub modulus: BigNumber,
+}
+
+#[derive(Serialize, Deserialize)]
+pub struct CSEncPrikey {
+    pub x1: Vec<BigNumber>,
+    pub x2: BigNumber,
+    pub x3: BigNumber,
+}
+
+#[derive(Serialize, Deserialize)]
+pub struct CSEncPubkey {
+    pub n: BigNumber,
+    pub two_inv_times_2: BigNumber, // (2^-1 % n) * 2, precomputation
+    pub paillier_group: PaillierGroup,
+    pub y1: Vec<BigNumber>,
+    pub y2: BigNumber,
+    pub y3: BigNumber,
+}
+
+#[derive(Serialize, Deserialize)]
+pub struct CSCiphertext {
+    pub u: BigNumber,
+    pub e: Vec<BigNumber>,
+    pub v: BigNumber,
+}
+
+pub struct CSKeypair {
+    pub pri_key: CSEncPrikey,
+    pub pub_key: CSEncPubkey,
+}
+
+impl PaillierGroup {
+    /// Order (modulus) is n^2
+    pub fn new(n: &BigNumber, ctx: &mut BigNumberContext) -> UrsaCryptoResult<Self> {
+        let modulus = n.sqr(Some(ctx))?; // n^2
+        let mut n_mul_2 = n.try_clone()?; // n*2
+        n_mul_2 = n_mul_2.lshift1()?;
+        let g_prime = modulus.rand_range()?;
+        let g = g_prime.mod_exp(&n_mul_2, &modulus, Some(ctx))?;
+        Ok(Self {
+            g,
+            h: n.increment()?,    // h = n+1
+            n_by_4: n.rshift(2)?, // n/4
+            modulus,
+        })
+    }
+
+    /// self.g^exp % self.modulus
+    pub fn exponentiate_g(
```

Comment:
Bignumber has the name `mod_exp`. Can we rename to `g_mod_exp`?

Reply:
`mod` is intentionally skipped. Since the methods are defined in the `PaillierGroup` object, its obvious that the group operation will involve modulus. `Bignumber` is meant to be generic so it should say which operations involve a modulus and which do not.